### ISSUE TITLE
Make ITenantInfo's Id type a generic

### ISSUE THREAD

**Issue by julesrx** [Finbuckle/Finbuckle.MultiTenant#602]
Change ITenantInfo's Id type to a generic to allow for more flexibility when implementing the library. This would help since here, even if the Id is a number in the tenant store or the database, it needs to be converted into a string.

Essentially, change this :
```cs
public interface ITenantInfo
{
    string? Id { get; set; }
}
```

To this :
```cs
public interface ITenantInfo<TKey>
{
    TKey? Id { get; set; }
}
```

And keep the default TenantInfo to a string :
```cs
public class TenantInfo : ITenantInfo<string>
```

This is a big breaking change though, with lots of refactoring to do 😐.
(realized this could be a duplicate of #365)

**Comment by AndrewTriesToCode** [comment]
Hi, yes this on my todo list. And I plan to tackle it soonish.

**Comment by domn1995** [comment]
Should the interface be this instead? Feels wrong to force it to be nullable as in the OP.

```cs
public interface ITenantInfo<TId>
{
    TId Id { get; set; }
}
```

You would then use it like this for compatibility:

```cs
public class TenantInfo : ITenantInfo<string?>
```

Alternatively, you could also force the ID to be immutable, and never null (as I would personally expect) and have a huge refactor 😅

```cs
public interface ITenantInfo<TId>
    where TId : notnull
{
    TId Id { get; init; }
}
```

**Comment by hbulens** [comment]
It would be a great improvement if we could change the type to something other than a string. I've been shouted at today (not really, but almost) that the `TenantId` is a string in our SQL database. I'm inclined to agree, so I was wondering if there are any plans to work on this soon?

**Comment by AndrewTriesToCode** [comment]
I am working on it but it's nontrivial and will cause breaking changes because some areas assume a string... great thinking on my part years ago. I do plan to include this soon. I'm wrapping it up with a few other breaking changes for a single major release bump.

**Comment by shawndewet** [comment]
Just following up on the plans to implement this?  It's been preventing me from using Finbuckle to its full potential for years.

**Comment by AndrewTriesToCode** [comment]
Ive been experimenting on how to best approach it. Is the main issue for you on the runtime side or mainly with respect to the database and not wanting to store strings there?

**Comment by shawndewet** [comment]
For me the issue is that my database has `int` as the tenant Id (and in a db column named 'ProfileId').  However, I noted a possible solution to this in related Issue #365 which I am trying out...will post back here if I get it to work for me.

**Comment by mvanput** [comment]
@AndrewTriesToCode is this something that is still on the planning or for grab ups for pr? I'm willing to work on it.

**Comment by AndrewTriesToCode** [comment]
I have a bunch of other refactors in progress but I did have a heavy session with codex on this last night. It is a really deep change with a large ripple effect. I’d welcome some help with it, try to base off the next branch.

**Comment by mvanput** [comment]
I will work based on the next branch. The identifier remains a string or is there a case/funding to make that also generic?

**Comment by AndrewTriesToCode** [comment]
Haven’t had any requests regarding identifier and hadn’t planned  on tackling that but I’m not opposed to it either.

A lot has changed in next branch and the changelog isn’t updated (updates automatically at release) but the docs are updated in general and I’m sure you’ll figure it out. I’m curious for any feedback you have on those changes too.

**Comment by mvanput** [comment]
Agreed to keep that as is! Will work on it these days. Keep you posted

**Comment by AndrewTriesToCode** [comment]
@mvput just a heads up I am reworking a lot of code currently in the `next` branch. I'll keep it in place and do a `next2` for my purposes then later incorporate whatever you come up with. If you haven't done much yet I advise you work off `next2`

**Comment by mvanput** [comment]
@AndrewTriesToCode the changes you made in the `next2` branch are looking nice. How far are you in terms of reaching a 'final' design. I think it's best to add the generic tenant info when there are no major reworks in the branch. Nice to see a cache implementation in this version.

I did some prep work in advance, so happy to move forward with this! By introducing generic type for `ITenantInfo`, I think some non generic interfaces can be removed and go fully generic type. For example you have `ITenantContext` and `ITenantContext<TTenantInfo>`. With the change it becomes either `ITenantContext<TId>` or `ITenantContext<TTenantInfo, TId>`. Passing the generic type id, make me wonder if there is a use case for the non-generic tenant interface? 

My proposal would be to ditch the non-generic type interfaces/classes. What do you think about this?

**Comment by AndrewTriesToCode** [comment]
hello, I'm sorry for all the force pushes to next2 -- I don't plan any more for a while. I had similar thoughts and one thing I think might really help here is the use of incremental source generators. I haven't tested the idea much but I was thinking the library would define everything just to ITenantInfo and a source generator could create extension methods dynamically for the tenant type. What I'm not sure if the EFCore stuff, it might still require  generics. What do you think of this idea?

**Comment by mvanput** [comment]
No problem and thanks for all the nice (re)work done in this lib.

My thoughts were also on source generators and I think ef core will also benefit on this part, but requires some work to do this.

Maybe a first step is to introduce full generic type support and then discuss the incremental source generators part?

**Comment by AndrewTriesToCode** [comment]
I think `ITenantInfo<TId>` and `ITenantContext<TId>` makes a lot of sense and then source generators to overload methods where having strong typing would be beneficial. I'll let you cook and take a look when you have something to share. If you think you might not have the time to look into this in time for .NET 11 release then let me know and I'll prioritize it to. Thanks!

**Comment by mvanput** [comment]
Cool! I will work it out this way. I think there is enough time for the .NET 11 release. My planning is to have a working generic type these days. Then look at the source generators part.

**Comment by mvanput** [comment]
@AndrewTriesToCode the projects are updated to the generic type id. The web sample works. I need to update the test projects and the identity app sample.

You already can take a peek here https://github.com/mvanput/Finbuckle.MultiTenant/tree/next2

**Comment by AndrewTriesToCode** [comment]
Looks like you are making good progress. What difficulties have you run into? I thought removing `TTenantType` first via source generation might have been needed first to make this easier... but you proved that wrong. If you want to just get the Id working and leave the TTenantType removal to me that works too -- I'm not 100% I even want to go that direction. But `TId` is a welcome addition. Are you using AI to help with the refactor?

**Comment by mvanput** [comment]
No difficulties so far, except just a bunch of refactoring and a lot of non generic methods/classes that becomes generic. For the source generation I'm going to create a PoC to see how that works and if we want to follow up that approach. First I'm going to finish and polish this.

No AI is used for this refactor.

**Comment by mvanput** [comment]
An update on the progress, tests are all done and passing, updating the last things and probably will create a PR for this change this week

**Comment by AndrewTriesToCode** [comment]
It's coming along nicely. Looking forward to the PR. I will likely tweak a few things but overall this is great work.
